proof.go

```diff
@@ -222,64 +241,87 @@ func (proof Proof) verifyLeafHashes(nth *Hasher, verifyCompleteness bool, nID na
-			return popIfNonEmpty(&proof.nodes)
+			return popIfNonEmpty(&proof.nodes), nil
 		}
 
 		// Recursively get left and right subtree
 		k := getSplitPoint(end - start)
-		left := computeRoot(start, start+k)
-		right := computeRoot(start+k, end)
+		left, err := computeRoot(start, start+k)
+		if err != nil {
+			return nil, fmt.Errorf("failed to compute subtree root [%d, %d): %w", start, start+k, err)
+		}
+		right, err := computeRoot(start+k, end)
+		if err != nil {
+			return nil, fmt.Errorf("failed to compute subtree root [%d, %d): %w", start+k, end, err)
+		}
 
 		// only right leaf/subtree can be non-existent
 		if right == nil {
-			return left
+			return left, nil
+		}
+		hash, err := nth.HashNode(left, right)
+		if err != nil {
+			return nil, fmt.Errorf("failed to hash node: %w", err)
 		}
-		hash := nth.HashNode(left, right)
-		return hash
+		return hash, nil
 	}
 
 	// estimate the leaf size of the subtree containing the proof range
 	proofRangeSubtreeEstimate := getSplitPoint(proof.end) * 2
 	if proofRangeSubtreeEstimate < 1 {
 		proofRangeSubtreeEstimate = 1
 	}
-	rootHash := computeRoot(0, proofRangeSubtreeEstimate)
+	rootHash, err := computeRoot(0, proofRangeSubtreeEstimate)
+	if err != nil {
+		return false, fmt.Errorf("failed to compute root [%d, %d): %w", 0, proofRangeSubtreeEstimate, err)
+	}
 	for i := 0; i < len(proof.nodes); i++ {
-		rootHash = nth.HashNode(rootHash, proof.nodes[i])
+		rootHash, err = nth.HashNode(rootHash, proof.nodes[i])
+		if err != nil {
+			return false, fmt.Errorf("failed to hash node: %w", err)
+		}
 	}
 
-	return bytes.Equal(rootHash, root)
+	return bytes.Equal(rootHash, root), nil
 }
 
 // VerifyInclusion checks that the inclusion proof is valid by using leaf data
-// and the provided proof to regenerate and compare the root. Note that the leaf
-// data should not contain the prefixed namespace, unlike the tree.Push method,
+// and the provided proof to regenerate and compare the root. Note that the leavesWithoutNamespace data should not contain the prefixed namespace, unlike the tree.Push method,
 // which takes prefixed data. All leaves implicitly have the same namespace ID:
 // `nid`.
-func (proof Proof) VerifyInclusion(h hash.Hash, nid namespace.ID, leaves [][]byte, root []byte) bool {
+func (proof Proof) VerifyInclusion(h hash.Hash, nid namespace.ID, leavesWithoutNamespace [][]byte, root []byte) bool {
 	nth := NewNmtHasher(h, nid.Size(), proof.isMaxNamespaceIDIgnored)
-	hashes := make([][]byte, len(leaves))
-	for i, d := range leaves {
+	hashes := make([][]byte, len(leavesWithoutNamespace))
+	for i, d := range leavesWithoutNamespace {
+		// prepend the namespace to the leaf data
 		leafData := append(
 			append(make([]byte, 0, len(d)+len(nid)), nid...), d...,
 		)
-		hashes[i] = nth.HashLeaf(leafData)
+		res, err := nth.HashLeaf(leafData)
+		if err != nil {
+			return false // this never can happen since the leafData is guaranteed to be namespaced
```

Comment:
[question] given this should never happen, should it `panic`?

Reply:
The same argument as in [this](https://github.com/celestiaorg/nmt/pull/136#discussion_r1145349472) also applies in here.